README.md
# chromecast-osx-audio

Streams macOS audio input to a local Chromecast device.

## Installation

To install the module for use in your projects:

```bash
npm install -g chromecast-osx-audio
```

## Usage

Global installation exposes the `chromecast` command to your shell. Running this
command will start listening to input, and connect to a local Chromecast with a
stream of that input.

To direct system audio, use [Soundflower](#soundflower).

```bash
$ chromecast --help

Usage: chromecast [options]

Options:
   -p, --port        The port that the streaming server will listen on.  [3000]
   -b, --bitrate     The bitrate for the mp3 encoded stream.  [192]
   -m, --mono        The stream defaults to stereo. Set to mono with this flag.
   -s, --samplerate  The sample rate for the mp3 encoded stream  [44100]
   -n, --name        A name for the server to report itself as.  [Chrome OSX Audio Stream]
   -u, --url         The relative URL that the stream will be hosted at.  [stream.mp3]
   -i, --iface       The public interface that should be reported. Selects the first interface by default.
   -l, --list        List devices available for streaming.
   -d, --device      Specify device to use for streaming.
   --version         print version and exit
```

### Soundflower

To direct system audio requires an additional piece of software. You see, macOS doesn’t provide native methods for accessing system audio (the output of the system mixer) so anything that’s not an input (like the microphone) is inaccessible. However, there’s a open source utility called [Soundflower][soundflower] that solves this for you. If you’re unfamiliar with it, here’s a step by step to get it working:

1. Install Soundflower. Reboot.
2. Install chromecast-osx-audio globally, with an `npm install -g chromecast-osx-audio`
3. Open your *System Preferences -> Sound* preference pane, and select “Soundflower (2ch)” as both your input and output.
4. Run `chromecast` in your terminal. It will find the first-available Chromecast, and stream your system audio to it.
5. Play your music/audio as normal. There is a 5–15 second delay due to how the Chromecast buffers.
6. Enjoy

## Environment Variables

None yet.

## Contributing

Feel free to send pull requests! I'm not picky, but would like the following:

1. Write tests for any new features, and do not break existing tests.
2. Be sure to point out any changes that break API.

## History

See [CHANGELOG](./CHANGELOG.md).

## The MIT License (MIT)

Copyright (c) 2014 Nathan Wittstock

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
the Software, and to permit persons to whom the Software is furnished to do so,
subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

[soundflower]: https://github.com/mattingalls/Soundflower


CHANGELOG.md
# Changelog

- **v1.0.1**  
README updates and clarifications [@hugovk](https://github.com/hugovk)

- **v1.0.0**  
Compatible with the latest versions of Node.js, thanks to a contribution to the
upstream [osx-audio][] package by [Alexander Kuznetsov][alexkuz].

- **v0.3.0**  
Adds ability to output list of devices as JSON (cli flag `-j`); refactors device
search. Updates castv2-client ([@danjenkins](https://github.com/fardog/node-chromecast-osx-audio/pull/9))

- **v0.2.0**  
Adds device listing and ability to specify a device if more than one are
found ([@pixelcat](https://github.com/fardog/node-chromecast-osx-audio/pull/2))

- **v0.1.3**  
Adds simple error handling.

- **v0.1.2**  
Updates webcast-osx-audio dependency.

- **v0.1.1**  
Updates webcast-osx-audio dependency.

- **v0.1.0**  
Splits webcast functionality to the [webcast-osx-audio](https://github.com/fardog/node-webcast-osx-audio) package.

- **v0.0.3**  
Updates to latest version of osx-audio.

- **v0.0.2**  
Readme was all wrong. Republished.

- **v0.0.1**  
Initial Release.

[osx-audio]: https://www.npmjs.com/package/osx-audio
[alexkuz]: https://github.com/alexkuz


lib/index.js
var Webcast = require('webcast-osx-audio');
var Client = require('castv2-client').Client;
var DefaultMediaReceiver = require('castv2-client').DefaultMediaReceiver;
var chromecast = require('./chromecast');

var Chromecast = function(options) {
  var deviceFound = false;
  
  if (options['l']) {
    console.log("Devices available for casting:");
  }

  chromecast.search(function (err, devices) {
    if (err) {

      return;
    }

    if (devices) {
      if (options['j']){
        console.log(JSON.stringify(devices));
        return;
      }

      Object.keys(devices).forEach(function (key) {
        var device = devices[key];
        if (typeof options['d'] !== 'undefined') {
          if (options['d'] == device.name) {
            // only start the service for the specified device.
            ondeviceup(device.name, device.addresses[0]);
          }
        }
        else {
          if (!options['l']) {
            console.log('found device "%s" at %s:%d', device.name, device.addresses[0], device.port);
          }
          // -d wasn't specified. Start for the first specified
          ondeviceup(device.name, device.addresses[0]);
        }
      })
    }

  });


  function ondeviceup(name, host) {
    if (options['l']) {
      console.log(name);
      return;
    }
    // don't try and start more than one connection (will fail if multiple devices are found)
    if (deviceFound) {
      return;
    }
    deviceFound = true;

    console.log("Connecting to device '%s'", name);
    var client = new Client();

    client.connect(host, function() {
      console.log('connected, launching app ...');

      var webcast = new Webcast(options);

      client.launch(DefaultMediaReceiver, function(err, player) {
        if (err) {
          console.error('Error: %s', err.message);
          client.close();
          return;
        }
        var media = {

          contentId: 'http://' + webcast.ip + ':' + options.port + '/' + options.url,
          contentType: 'audio/mpeg3',
          streamType: 'LIVE', // or LIVE

          // Title and cover displayed while buffering
          metadata: {
            type: 0,
            metadataType: 0,
            title: options.name 
          }        
        };

        player.on('status', function(status) {
          console.log('status broadcast playerState=%s', status.playerState);
        });

        console.log('app "%s" launched, loading media %s ...', player.session.displayName, media.contentId);

        player.load(media, { autoplay: true }, function(err, status) {
          if (err) {
            console.error('Error: %s', err.message);
            client.close();
            return;
          }
          console.log('media loaded playerState=%s', status.playerState);
        });
      });

    });

    client.on('error', function(err) {
      console.error('Error: %s', err.message);
      client.close();
    });

  }
}

module.exports = Chromecast;


lib/chromecast.js
var mdns = require('mdns');

var devices = {};
var browser = mdns.createBrowser(mdns.tcp('googlecast'));

var chromecast = {
  start: function(){
    browser.start();
  },

  stop: function () {
    browser.stop();
  },

  stream: function (cb) {
    browser.on('serviceUp', function (service) {
      devices[service.name] = service;
      cb && cb(null, services);
    });

    browser.on('serviceDown', function (service){
      delete devices[service.name];
      cb && cb(null, services);
    });

    this.start();
  },
  search: function (cb) {
    var numDevices = 0;
    this.stream();
    var interval = setInterval(function () {
      if (numDevices === Object.keys(devices).length) {
        clearInterval(interval);
        chromecast.stop();
        cb(null, devices);
      }
      numDevices = Object.keys(devices).length;
    }, 20);
  }
};

module.exports = chromecast;


lib/cli.js
'use strict';

var fs = require('fs');
var chalk = require('chalk');
var parseArgs = require('minimist');

// Number.isInteger() polyfill ::
// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isInteger
if (!Number.isInteger) {
  Number.isInteger = function isInteger (nVal) {
    return typeof nVal === "number" && isFinite(nVal) && nVal > -9007199254740992 && nVal < 9007199254740992 && Math.floor(nVal) === nVal;
  };
}

var cli = function(options) {
  this.options = {
    alias: {
      port: 'p',
      bitrate: 'b',
      mono: 'm',
      samplerate: 's',
      name: 'n',
      url: 'u',
      iface: 'i',
      version: 'v',
      help: 'h',
      device: 'd',
      list: 'l',
      json: 'j'
    },
    default: {
      port: 3000,
      bitrate: 192,
      samplerate: 44100,
      name: 'Chrome OSX Audio Stream',
      url: 'stream.mp3'
    },
    'boolean': ['version', 'help', 'mono'],
    'string': ['name'],
    'integer': ['port', 'bitrate', 'samplerate']
  };

  this.errors = [];
  this.message = null;

  this.helpMessage = [
    chalk.bold.blue("Usage: chromecast [options]"),
    "",
    "Options:",
    "   -p, --port        The port that the streaming server will listen on.  [3000]",
    "   -b, --bitrate     The bitrate for the mp3 encoded stream.  [192]",
    "   -m, --mono        The stream defaults to stereo. Set to mono with this flag.",
    "   -s, --samplerate  The sample rate for the mp3 encoded stream  [44100]",
    "   -n, --name        A name for the server to report itself as.  [Chrome OSX Audio Stream]",
    "   -u, --url         The relative URL that the stream will be hosted at.  [stream.mp3]",
    "   -i, --iface       The public interface that should be reported. Selects the first interface by default.",
    "   -l, --list        List devices available for streaming.",
    "   -j, --json        List devices available for streaming as JSON.",
    "   -d, --device      Specify device to use for streaming.",
    "   --version         print version and exit"
  ];

  return this;
};

cli.prototype.parse = function(argv, next) {
  var options = parseArgs(argv, this.options);

  if (options.version) {
    var pkg = require('../package.json');
    this.message = "version " + pkg.version;
  }
  else if (options.help) {
    this.message = this.helpMessage.join('\n');
  }
  else {
    /*
     * Options are processed in a significant order; we only save the last error
     * message, so we'll want to make sure the most significant are last
     */

    // ensure that parameter-expecting options have parameters
    this.options['string'].forEach(function(i) {
       if(typeof options[i] !== 'undefined') {
         if (typeof options[i] !== 'string' || options[i].length < 1) {
           this.errors.push(new Error(i + " expects a value."));
         }
       }
    }.bind(this));

    // ensure that number-expecting options have parameters
    this.options['integer'].forEach(function(i) {
      if(typeof options[i] !== 'undefined') {
        if (!Number.isInteger(options[i])) {
          this.errors.push(new Error(i + " expects an integer value."));
        }
      }
    }.bind(this));
  }
  
  this.parsedOptions = options;
  
  if (typeof next === 'function') {
    // we return the array of errors if there are any, otherwise null
    next(this.errors.length > 0 ? this.errors : null, this.message, options);
  }

  return this;
};


module.exports = cli;
